framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java

```diff
@@ -894,7 +892,11 @@ public AnnotationFormatter getAnnotationFormatter() {
      *         empty set if no qualifiers are supported
      */
     public final Set<Class<? extends Annotation>> getSupportedTypeQualifiers() {
-        return supportedQuals;
+        if (this.supportedQuals.isEmpty()) {
+            supportedQuals.addAll(createSupportedTypeQualifiers());
```

Comment:
The origin contract requires sub-classes to return an immutable set in `javadoc`, but it cannot prevent the subclass actually return a mutable set, like this implementation in current Checker Framework Inference master branch:
https://github.com/typetools/checker-framework-inference/blob/master/src/nninf/NninfAnnotatedTypeFactory.java#L58

Thus, I think remove the requirement of return immutable set from subclasses, and let final method `getSupportedQualifiers()` always return an immutable view of `supportedQualifiers` would be safer.